File: leetcode/329_Longest_Increasing_Path_in_a_Matrix.py
from typing import List
from unittest import TestCase

class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        n, m = len(matrix), len(matrix[0])
        dp = [[0] * m for i in range(n)]

        def rec(i, j, dp):
            if dp[i][j] == 0:
                max_next_lengh = 0
                for i1, j1 in ((i + 1, j), (i, j + 1), (i - 1, j), (i, j - 1)):
                    if 0 <= i1 < n and 0 <= j1 < m and matrix[i][j] < matrix[i1][j1]:
                        next_lengh = rec(i1, j1, dp)
                        if next_lengh > max_next_lengh:
                            max_next_lengh = next_lengh
                dp[i][j] = max_next_lengh + 1
            return dp[i][j]

        max_length = 0
        for i in range(n):
            for j in range(m):
                if dp[i][j] == 0:
                    rec(i, j, dp)
                if dp[i][j] > max_length:
                    max_length = dp[i][j]
        return max_length

class LongestIncreasingPathTest(TestCase):
    def test_1(self):
        self.assertEqual(4, Solution().longestIncreasingPath([[9,9,4],[6,6,8],[2,1,1]]))

    def test_2(self):
        self.assertEqual(4, Solution().longestIncreasingPath([[3,4,5],[3,2,6],[2,2,1]]))        

    def test_3(self):
        self.assertEqual(1, Solution().longestIncreasingPath([[1]]))                

File: leetcode/329_Longest_Increasing_Path_in_a_Matrix.rs
struct Solution {}

impl Solution {
    pub fn longest_increasing_path(matrix: Vec<Vec<i32>>) -> i32 {
        fn rec(i: usize, j: usize, n: usize, m: usize, matrix: &Vec<Vec<i32>>, dp: &mut Vec<Vec<i32>>) -> i32 {
            if dp[i][j] == 0 {
                let mut max_next_length = 0;
                for (i1, j1) in [(i as i32 + 1, j as i32), (i as i32, j as i32 + 1), (i as i32 - 1, j as i32), (i as i32, j as i32 - 1)] {
                    if 0 <= i1 && i1 < n as i32 && 0 <= j1 && j1 < m as i32 && matrix[i][j] < matrix[i1 as usize][j1 as usize] {
                        max_next_length = max_next_length.max(rec(i1 as usize, j1 as usize, n, m, matrix, dp));
                    }
                }
                dp[i][j] = max_next_length + 1;
            }
            dp[i][j]
        }

        let (n, m) = (matrix.len(), matrix[0].len());
        let mut max_length = 0;
        let mut dp = vec![vec![0; m]; n];
        for i in 0..n {
            for j in 0..m {
                max_length = max_length.max(rec(i, j, n, m, &matrix, &mut dp));
            }
        }
        max_length
    }
}

#[test]
fn test_1() {
    assert_eq!(4, Solution::longest_increasing_path(vec![vec![9,9,4],vec![6,6,8],vec![2,1,1]]));
}

#[test]
fn test_2() {
    assert_eq!(4, Solution::longest_increasing_path(vec![vec![3,4,5],vec![3,2,6],vec![2,2,1]]));
}

#[test]
fn test_3() {
    assert_eq!(1, Solution::longest_increasing_path(vec![vec![1]]));
}